pkg/rpcclient/actor/waiter.go

```diff
@@ -34,37 +45,60 @@ type (
 	// RPCWSWaiter is an interface that enables improved transaction awaiting functionality
 	// for Actor instance based on web-socket Block and ApplicationLog notifications.
 	RPCWSWaiter interface {
-		RPCPollingWaiter
-
+		// Context should return the RPC client context to be able to gracefully
+		// shut down all running processes (if so).
+		Context() context.Context
 		SubscribeForNewBlocksWithChan(primary *int, since *uint32, rcvrCh chan<- rpcclient.Notification) (string, error)
 		SubscribeForTransactionExecutionsWithChan(state *string, container *util.Uint256, rcvrCh chan<- rpcclient.Notification) (string, error)
 		Unsubscribe(id string) error
 	}
 )
 
-// Wait allows to wait until transaction will be accepted to the chain. It can be
-// used as a wrapper for Send or SignAndSend and accepts transaction hash,
-// ValidUntilBlock value and an error. It returns transaction execution result
-// or an error if transaction wasn't accepted to the chain.
-func (a *Actor) Wait(h util.Uint256, vub uint32, err error) (*state.AppExecResult, error) {
-	if err != nil {
-		return nil, err
+// pollingWaiter is a polling-based Waiter.
+type pollingWaiter struct {
+	RPCPollingWaiter
+	version *result.Version
+}
+
+// wsWaiter is a websocket-based Waiter.
+type wsWaiter struct {
+	RPCWSWaiter
+	polling Waiter
+}
+
+// NewWaiter creates Waiter instance. It can be either websocket-based or
+// polling-base, otherwise an error is returned.
+func NewWaiter(ra RPCActor, version *result.Version) (Waiter, error) {
+	ws, wsOK := ra.(RPCWSWaiter)
+	polling, pollingOK := ra.(RPCPollingWaiter)
+	if wsOK && pollingOK {
+		return &wsWaiter{
+			RPCWSWaiter: ws,
+			polling: &pollingWaiter{
+				RPCPollingWaiter: polling,
+				version:          version,
+			},
+		}, nil
 	}
-	if wsW, ok := a.client.(RPCWSWaiter); ok {
-		return a.waitWithWSWaiter(wsW, h, vub)
+	if pollingOK {
+		return &pollingWaiter{
+			RPCPollingWaiter: polling,
+			version:          version,
+		}, nil
 	}
-	return a.waitWithSimpleWaiter(a.client, h, vub)
+	return nil, fmt.Errorf("%w: actor doesn't implement RPCPollingWaiter or RPCWSWaiter interface", ErrWaiterInit)
```

Comment:
I know practically we'll always have it with RPC clients, but it can be somewhat unobvious for alternative RPCActor implementations why they can't work for Actor. So we better let them work, just have a `NullWaiter`.